### PR TITLE
Pancan PREQUAL and new workflow STUDY_REDIRECT support  

### PR DESCRIPTION
**Pull request by ssettipalli**
PR covers below tickets:
https://broadinstitute.atlassian.net/browse/DDP-6324
https://broadinstitute.atlassian.net/browse/DDP-6313
https://broadinstitute.atlassian.net/browse/DDP-6326
https://broadinstitute.atlassian.net/browse/DDP-6328

For 6313 completion, we need couple of PEX related tickets (DDP-6322 and DDP-6323) to be implemented.
Will be revisiting and completing 6313 in a different PR
Couple of minor issues related to PEX and API Doc changes are still pending..

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java [hunk 1]
```diff
@@ -519,12 +519,19 @@ private Object applyAnswerPredicate(InterpreterContext ictx, String userGuid, St
...
         } else {
             var answerDao = new AnswerCachedDao(ictx.getHandle());
-            Optional<Answer> compositeAnswer;
             if (instanceGuid == null) {
-                compositeAnswer = answerDao.findAnswerByLatestInstanceAndQuestionStableId(userGuid, studyId, stableId);
-            } else {
-                compositeAnswer = answerDao.findAnswerByInstanceGuidAndQuestionStableId(instanceGuid, stableId);
+                //try to find latest activity instance so that if composite answers exists can load child answers
+                instanceGuid = ictx.getHandle().attach(ActivityInstanceDao.class)
+                        .findLatestActivityInstanceGuidByUserStudyAndQuestionStableId(userGuid, studyId, stableId);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/pex/TreeWalkInterpreter.java]: I think finding latest activity instance using question stable id is a little weird. We have the `activityCode`. Can we use that instead? I think it'll make more sense that way.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowStudyRedirectState.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Optional;
+
+import org.broadinstitute.ddp.model.workflow.StudyRedirectState;
+import org.jdbi.v3.sqlobject.SqlObject;
+import org.jdbi.v3.sqlobject.config.RegisterConstructorMapper;
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+
+public interface JdbiWorkflowStudyRedirectState extends SqlObject {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowStudyRedirectState.java]: Can we put these into `JdbiWorkflowState` so we have one less file to deal with? Even better, merge and remove `JdbiWorkflowActivityState` as well.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowStudyRedirectState.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    @SqlUpdate("insert into workflow_study_redirect_state (workflow_state_id, study_guid, redirect_url) "
+            + " values (:stateId, :studyGuid, :redirectUrl)")
+    int insert(@Bind("stateId") long stateId, @Bind("studyGuid") String studyGuid, @Bind("redirectUrl") String redirectUrl);
+
+    @SqlQuery("select workflow_state_id, study_guid, redirect_url from workflow_study_redirect_state "
+            + " where study_guid = :studyGuid and redirect_url = :redirectUrl")
+    @RegisterConstructorMapper(StudyRedirectState.class)
+    Optional<StudyRedirectState> findIdByStudyNameGuidAndRedirectUrl(@Bind("studyGuid") String studyGuid,
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/JdbiWorkflowStudyRedirectState.java]: Looks like we're returning an entire object and not just id.
```suggestion
    Optional<StudyRedirectState> findByStudyNameGuidAndRedirectUrl(@Bind("studyGuid") String studyGuid,
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/workflow/StaticState.java [hunk 4]
```diff
@@ -32,6 +32,10 @@ public static StaticState participantList() {
         return new StaticState(StateType.PARTICIPANT_LIST);
     }
 
+    public static StaticState studyRedirect() {
+        return new StaticState(StateType.STUDY_REDIRECT);
+    }
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/workflow/StaticState.java]: `study_redirect` is not a "static" state (e.g. it has associated data and not just a type). It does not belong here.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/json/workflow/WorkflowResponse.java [hunk 5]
```diff
@@ -8,6 +8,16 @@
...
     protected String next;
 
+    @SerializedName("studyName")
+    protected String studyName;
+
+    @SerializedName("studyGuid")
+    protected String studyGuid;
+
+    @SerializedName("redirectUrl")
+    protected String redirectUrl;
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/json/workflow/WorkflowResponse.java]: Can we put these into a subclass?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/workflow/StudyRedirectState.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package org.broadinstitute.ddp.model.workflow;
+
+import org.jdbi.v3.core.mapper.reflect.ColumnName;
+import org.jdbi.v3.core.mapper.reflect.JdbiConstructor;
+
+public class StudyRedirectState implements WorkflowState {
+
+    Long workflowStateId;
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/workflow/StudyRedirectState.java]: ID should not be null.

File: pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+
+    <changeSet author="sampath" id="20210711-workflow-study-redirect-state-table">
+        <createTable tableName="workflow_study_redirect_state">
+            <column name="workflow_state_id" type="bigint">
+                <constraints primaryKey="true" primaryKeyName="workflow_activity_state_pk"
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml]:
```suggestion
                <constraints primaryKey="true" primaryKeyName="workflow_study_redirect_state_pk"
```

File: pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                             references="workflow_state(workflow_state_id)" foreignKeyName="workflow_study_redirect_state_workflow_state_fk"/>
+            </column>
+            <column name="study_guid" type="varchar2(10))">
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml]: I think `varchar2` is a data type in Oracle but not MySQL.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg [hunk 9]
```diff
@@ -1,15 +1,21 @@
 group WorkflowDao;
 
 queryOrderedActiveNextStateCandidatesByStudyIdAndFromStateId() ::= <<
-select trans.workflow_transition_id,
+select trans.workflow_transition_id, trans.from_state_id,
        next_state_type.workflow_state_type_code,
        next_act_state.study_activity_id,
+       next_redirect_state.study_guid,
+       next_redirect_state.redirect_url,
+       us_redirect.study_name,
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: I think we might be better off getting it from the `StudyDto` instead. We're passing in `studyGuid` in `WorkflowService` but we should probably change that.

File: study-builder/config/vars.conf.ctmpl [hunk 10]
```diff
@@ -25,6 +25,20 @@
     "sendgridDefaultSalutation" : "{{$conf.sendgridDefaultSalutation}}",
     "analyticsToken": "{{$conf.analyticsToken}}",
 
+{{if eq $study "pancan"}}
+    "angioStudyGuid": "{{$conf.angioStudyGuid}}",
```
Comment (yufengwng) [inline study-builder/config/vars.conf.ctmpl]: Study guid doesn't change, so don't need to put it into Vault...

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java [hunk 11]
```diff
@@ -77,6 +78,10 @@ private WorkflowState convertWorkflowState(Handle handle, Config stateCfg) {
             String activityCode = stateCfg.getString("activityCode");
             long activityId = ActivityBuilder.findActivityId(handle, studyDto.getId(), activityCode);
             return new ActivityState(activityId);
+        } else if (type == StateType.STUDY_REDIRECT) {
+            String studyGuid = stateCfg.getString("studyGuid");
+            String redirectUrl = stateCfg.getString("redirectUrl");
+            return new StudyRedirectState(studyGuid, redirectUrl);
```
Comment (yufengwng) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/WorkflowBuilder.java:84]: Can you update `stateAsStr()` as well?

File: study-builder/studies/pancan/i18n/en.conf [hunk 12]
```diff
@@ -17,21 +17,179 @@
 "prequal_sections_01_question_01_picklistOptions_02_optionLabelTemplate"="My child has been diagnosed with cancer and signing up with them or for them"
 "prequal_sections_01_question_01_validations_hintTemplate"="Please choose one of the above options."
 
-"prequal_diagnosed_primary_cancers_prompt"="What primary cancer(s) have you/your child been diagnosed with?"
-"prequal_sections_02_question_01_validations_01_hintTemplate_01"="If you/your child have been diagnosed with more than one cancer type, please select all that apply."
+"prequal_diagnosed_primary_cancer_list_prompt"="What primary cancer(s) have you/your child been diagnosed with? \n If you/your child have been diagnosed with more than one cancer type, please use the button below to add an additional diagnosis."
+"prequal_sections_02_question_01_validations_01_hintTemplate_01"="If you/your child have been diagnosed with more than one cancer type, please use the button below to add an additional diagnosis."
```
Comment (yufengwng) [inline study-builder/studies/pancan/i18n/en.conf]: Can you cleanup these variable names and use JSON-like syntax so it's more consistent?

File: study-builder/studies/pancan/i18n/en.conf [hunk 13]
```diff
@@ -17,21 +17,179 @@
 "prequal_sections_01_question_01_picklistOptions_02_optionLabelTemplate"="My child has been diagnosed with cancer and signing up with them or for them"
 "prequal_sections_01_question_01_validations_hintTemplate"="Please choose one of the above options."
 
-"prequal_diagnosed_primary_cancers_prompt"="What primary cancer(s) have you/your child been diagnosed with?"
-"prequal_sections_02_question_01_validations_01_hintTemplate_01"="If you/your child have been diagnosed with more than one cancer type, please select all that apply."
+"prequal_diagnosed_primary_cancer_list_prompt"="What primary cancer(s) have you/your child been diagnosed with? \n If you/your child have been diagnosed with more than one cancer type, please use the button below to add an additional diagnosis."
```
Comment (yufengwng) [inline study-builder/studies/pancan/i18n/en.conf]: The prompt is going into HTML, so I don't think `\n` is what you want. Maybe try `<br/>`. (Have we considered a tooltip?)

File: study-builder/studies/pancan/prequal.conf [hunk 14]
```diff
@@ -310,20 +310,42 @@
               "shownExpr": null
             },
             {
-              "question": ${_includes.diagnosed_primary_cancers} {
-                "stableId": "DIAGNOSED_PRIMARY_CANCERS"
+              "question": ${_includes.diagnosed_primary_cancer_list} {
+                "stableId": "DIAGNOSED_PRIMARY_CANCER_LIST"
```
Comment (yufengwng) [inline study-builder/studies/pancan/prequal.conf]: Please pull out identifiers to `subs.conf`.

File: study-builder/studies/pancan/snippets/common/picklist-question-yes-no-notsure.conf [hunk 15]
```diff
@@ -0,0 +1,92 @@
+{
+  "questionType": "PICKLIST",
+  "selectMode": "SINGLE",
+  "renderMode": "LIST",
+  "isRestricted": false,
+  "isDeprecated": false,
+  "hideNumber": false,
+  "additionalInfoHeaderTemplate": null,
+  "additionalInfoFooterTemplate": null,
+  "validations": [],
```
Comment (yufengwng) [inline study-builder/studies/pancan/snippets/common/picklist-question-yes-no-notsure.conf:10]: Should these be required (if shown)?

File: study-builder/studies/pancan/snippets/question-diagnosed-primary-cancer.conf [hunk 16]
```diff
@@ -0,0 +1,2516 @@
...
+      {
+        "name": "prequal_diagnosed_primary_cancer_prompt",
+        "translations": [
+          { "language": "en", "text":${i18n.en.prequal_diagnosed_primary_cancer_prompt} },
+          { "language": "es", "text": ${i18n.es.prequal_diagnosed_primary_cancer_prompt} }
+        ]
+      }
+    ]
+  },
+  "picklistOptions": [
```
Comment (yufengwng) [inline study-builder/studies/pancan/snippets/question-diagnosed-primary-cancer.conf:17]: This is good for now, but let's make a note to come back and group these into nested options.

File: study-builder/studies/pancan/i18n/en.conf (hunk 13 above)
Comment (bskinner) [reply study-builder/studies/pancan/i18n/en.conf]: HTML elements are not permitted  in the question prompts on the mobile apps if the new line is important for the formatting of the question.

Comment (yufengwng) [reply study-builder/studies/pancan/i18n/en.conf]: HTML elements in question prompt will get stripped for clients that wants it trimmed down.

Comment (yufengwng) [reply study-builder/studies/pancan/i18n/en.conf]: I would have recommended "additional info header" but I don't think web client supports that at the moment.

Comment (bskinner) [reply study-builder/studies/pancan/i18n/en.conf]: Additional info header appears above the prompt in mobile.

Comment (bskinner) [reply study-builder/studies/pancan/i18n/en.conf]: > HTML elements in question prompt will get stripped for clients that wants it trimmed down.

Yes, but that doesn't address my point. If that newline is important to the presentation, switching to a `</br>` will result in everything being squashed onto a single wrapped line.

Comment (yufengwng) [reply study-builder/studies/pancan/i18n/en.conf]: My guess is that whether it's shown on one line or two lines is not that important. And if it was, we (study staff) should account for the limitations on non-web clients (such as mobile).

Comment (bskinner) [reply study-builder/studies/pancan/i18n/en.conf:20]: What about `picklistLabel`? In the mobile apps, this is currently presented as a label placed below the prompt, using a lighter grey coloration & slight smaller font. Given the text here is ancillary information, that may be a good place for it.

Comment (bskinner) [reply study-builder/studies/pancan/i18n/en.conf]: What about `picklistLabel`? In the mobile apps, this is currently presented as a label placed below the prompt, using a lighter grey coloration & slight smaller font. Given the text here is ancillary information, that may be a good place for it, if the presentation in the Angular client is compatible.

```json
{
  "prompt": "What primary cancer(s) have you/your child been diagnosed with?",
  "picklistLabel": "If you/your child have been diagnosed with more than one cancer type, please use the button below to add an additional diagnosis."
}
```

One other thing, is the you/your child wording correct? It reads like it should be "you have/your child has"- "your child have" doesn't fit in the context.

File: study-builder/studies/pancan/prequal.conf (hunk 14 above)
Comment (ssettipalli) [reply study-builder/studies/pancan/prequal.conf]: I sort of like question stableIds not in substitutions but refactored any way.

File: study-builder/studies/pancan/i18n/en.conf (hunk 13 above)
Comment (ssettipalli) [reply study-builder/studies/pancan/i18n/en.conf]: separated into two text content blocks.
Text will be either: you OR your child depending on who.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java [hunk 17]
```diff
@@ -75,6 +77,17 @@ default int deleteStudyWorkflow(long studyId) {
...
+            String studyGuid = ((StudyRedirectState) state).getStudyGuid();
+            String redirectUrl = ((StudyRedirectState) state).getRedirectUrl();
+            Optional<StudyRedirectState> redirectStateOpt =  getJdbiWorkflowState().findByStudyNameGuidAndRedirectUrl(
+                    studyGuid, redirectUrl);
+            if (redirectStateOpt.isPresent()) {
+                StudyRedirectState redirectState = redirectStateOpt.get();
+                return Optional.of(redirectState.getWorkflowStateId());
+            } else {
+                return Optional.empty();
+            }
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/WorkflowDao.java]: Next time, you can replace this whole thing with
```java
return getJdbiWorkflowState()
  .findByStudyNameGuidAndRedirectUrl(...)
  .map(StudyRedirectState::getWorkflowStateId);
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java [hunk 18]
```diff
@@ -89,6 +91,10 @@ public WorkflowResponse buildStateResponse(Handle handle, String userGuid, Workf
                 return handle.attach(WorkflowDao.class)
                         .findActivityCodeAndLatestInstanceGuidAsResponse(activityId, userGuid)
                         .orElseThrow(() -> new NoSuchElementException("Could not find activity data to build response for " + state));
+            } else if (state.getType() == StateType.STUDY_REDIRECT) {
+                StudyRedirectState studyRedirectState = (StudyRedirectState)state;
+                return new WorkflowStudyRedirectResponse(studyRedirectState.getStudyName(),
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java]: We should get the study name from `StudyI18nDto`.

File: pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                   http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.5.xsd">
+
+    <changeSet author="sampath" id="20210711-workflow-study-redirect-state-table">
+        <createTable tableName="workflow_study_redirect_state">
+            <column name="workflow_state_id" type="bigint">
+                <constraints primaryKey="true" primaryKeyName="workflow_study_redirect_state_pk"
+                             references="workflow_state(workflow_state_id)" foreignKeyName="workflow_study_redirect_state_workflow_state_fk"/>
+            </column>
+            <column name="study_guid" type="varchar(10))">
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-6313-add-study-redirect-workflow.xml]: There's an extra closing parenthesis here and below.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java [hunk 20]
```diff
@@ -89,13 +98,40 @@ public WorkflowResponse buildStateResponse(Handle handle, String userGuid, Workf
...
+        if (userProfile.isPresent()) {
+            preferredLanguageId = userProfile.get().getPreferredLangId();
+        }
+        if (preferredLanguageId == null) {
+            preferredLanguageId = LanguageStore.getDefault().getId();
+        }
+        Optional<StudyI18nDto> preferredTranslation = translationDao.findTranslationByStudyIdAndLanguageCodeId(
+                study.getId(),
+                preferredLanguageId
+        );
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java]: What if study didn't configure an `StudyI18n` for their preferred language? How about we get list of all the `StudyI18n`, and if their preferred language one is not in the list, we default to the study's default one (or English)?

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg [hunk 21]
```diff
@@ -4,12 +4,16 @@ queryOrderedActiveNextStateCandidatesByStudyIdAndFromStateId() ::= <<
...
+       next_redirect_state.study_guid,
+       next_redirect_state.redirect_url,
        expr.expression_text
   from workflow_transition as trans
   join workflow_state as next_state on next_state.workflow_state_id = trans.next_state_id
   join workflow_state_type as next_state_type on next_state_type.workflow_state_type_id = next_state.workflow_state_type_id
   left join workflow_activity_state as next_act_state on next_act_state.workflow_state_id = next_state.workflow_state_id
   left join expression as expr on expr.expression_id = trans.precondition_expression_id
+  left join workflow_study_redirect_state as next_redirect_state on next_redirect_state.workflow_state_id = next_state.workflow_state_id
+  join umbrella_study as us on us.umbrella_study_id = trans.umbrella_study_id
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: Don't need this `join` anymore.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java (hunk 20 above)
Comment (ssettipalli) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java]: considering the requirement is limited to existing cmi studies [angio, brain, osteo, mbc, mpc, esc], may be we are  over working on this studyName  translation! We have translations for just MBC I believe. 
Changed it any way.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg (hunk 21 above)
Comment (ssettipalli) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: We do need the join to workflow_study_redirect_state to get studyGuid and redirectUrl.

Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: D'oh. Oh right, I missed that.

Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: Wait no. I meant we don't need the `join` to umbrella_study. We can keep the `left join`.

Comment (ssettipalli) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/WorkflowDao.sql.stg]: Yeah removed that join

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java (hunk 20 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java]: I don't see any difference between new implementation and old... anyways, I guess it's fine. Just make a note to come back to it. You never know, some other study is going to use this feature and then we later find out that we don't do the right handling for language...

**Review by yufengwng — APPROVED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java (hunk 20 above)
Comment (ssettipalli) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java]: will revisit when I finish rest of pending prequal stuff

**Pull request merged**